File: ios/SyxVPNCore/Sources/SyxVPNCore/SessionStore.swift
import Foundation

/// Where the session's values are actually kept.
///
/// Behind a protocol for one practical reason and one design one. The practical
/// one: the Keychain needs an entitlement and a signed host application, so a
/// store nailed to it could not be tested by `swift test` at all. The design
/// one: the tunnel extension and the app are different processes reading the
/// same values, and which container they share is a decision for the app
/// target, not for this file.
public protocol SessionStorage: AnyObject {
    func string(_ key: String) -> String?
    func setString(_ key: String, _ value: String?)
    func bool(_ key: String, default fallback: Bool) -> Bool
    func setBool(_ key: String, _ value: Bool)
    func removeAll()
}

/// The keys whose values are bearer secrets.
///
/// A backend that can offer two levels of protection — Keychain for these,
/// ordinary preferences for the rest — reads this rather than guessing. The
/// subscription URL is on the list: on its own it is enough to use the account.
public enum SessionSecrets {
    public static let keys: Set<String> = ["token", "subscription_url", "profiles"]
}

/// A store with no persistence, for tests.
public final class InMemorySessionStorage: SessionStorage {
    private var values: [String: String] = [:]
    private var flags: [String: Bool] = [:]

    public init() {}

    public func string(_ key: String) -> String? { values[key] }
    public func setString(_ key: String, _ value: String?) {
        if let value { values[key] = value } else { values.removeValue(forKey: key) }
    }
    public func bool(_ key: String, default fallback: Bool) -> Bool { flags[key] ?? fallback }
    public func setBool(_ key: String, _ value: Bool) { flags[key] = value }
    public func removeAll() { values.removeAll(); flags.removeAll() }
}

/// Holds the customer session token, the last known subscription URL, and the
/// settings the person has chosen.
///
/// The token and the subscription URL are bearer secrets: whoever holds either
/// can use the account. Which store keeps them is the backend's business; that
/// they are named as secrets is this file's.
public final class SessionStore {

    private enum Key {
        static let token = "token"
        static let email = "email"
        static let subscriptionUrl = "subscription_url"
        static let profiles = "profiles"
        static let hidden = "hidden_configs"
        static let imported = "imported_configs"
        static let automatic = "automatic_server"
        static let country = "country"
        static let purpose = "purpose"
        static let memory = "connection_memory"
        static let controlBase = "control_plane_base"
        static let dns = "dns_mode"
    }

    private let storage: SessionStorage

    /// Whether this build offers encrypted DNS at all.
    ///
    /// It gates the mode outright rather than only setting the default: a build
    /// with the flag off has to behave exactly like the build before the
    /// feature, including for the phone of somebody who had already picked an
    /// encrypted mode. That is what makes it a switch that can be thrown back.
    private let privateDnsEnabled: Bool

    public init(storage: SessionStorage, privateDnsEnabled: Bool = true) {
        self.storage = storage
        self.privateDnsEnabled = privateDnsEnabled
    }

    public var token: String? {
        get { storage.string(Key.token) }
        set { storage.setString(Key.token, newValue) }
    }

    /// Whether there is a session at all. A fresh install has none, and that is
    /// fine — the tunnel does not need one.
    public var signedIn: Bool { token != nil }

    public var email: String? {
        get { storage.string(Key.email) }
        set { storage.setString(Key.email, newValue) }
    }

    public var subscriptionUrl: String? {
        get { storage.string(Key.subscriptionUrl) }
        set { storage.setString(Key.subscriptionUrl, newValue) }
    }

    /// Last profile list fetched from the subscription, newline separated.
    public var cachedProfiles: String? {
        get { storage.string(Key.profiles) }
        set { storage.setString(Key.profiles, newValue) }
    }

    /// Configs the person pasted in themselves, newline separated.
    ///
    /// These are not a subscription and are nobody's to revoke: a config bought
    /// from another provider belongs to whoever bought it, so it survives
    /// signing out, an expired plan, and never having had one. Only removing it
    /// here removes it.
    public var importedConfigs: String? {
        get { storage.string(Key.imported) }
        set { storage.setString(Key.imported, newValue) }
    }

    /// The pasted configs as lines, in the order they were added.
    public var importedLines: [String] {
        (importedConfigs ?? "")
            .split(omittingEmptySubsequences: true, whereSeparator: \.isNewline)
            .map { $0.trimmed }
            .filter { !$0.isEmpty }
    }

    public func addImported(_ uris: [String]) {
        if uris.isEmpty { return }
        importedConfigs = (importedLines + uris).joined(separator: "\n")
    }

    public func removeImported(_ uri: String) {
        let kept = importedLines.filter { $0 != uri }.joined(separator: "\n")
        importedConfigs = kept.isEmpty ? nil : kept
    }

    /// Servers the person chose to hide, as "host:port" entries.
    ///
    /// A subscription decides which servers exist, so hiding is local and
    /// sticky: a refresh brings the server back from the control plane but it
    /// stays out of the list until it is unhidden here.
    public var hiddenConfigs: Set<String> {
        get {
            Set((storage.string(Key.hidden) ?? "")
                .split(separator: "\n", omittingEmptySubsequences: true)
                .map(String.init))
        }
        set {
            // Sorted, so the same set always writes the same bytes.
            storage.setString(
                Key.hidden, newValue.isEmpty ? nil : newValue.sorted().joined(separator: "\n")
            )
        }
    }

    /// Which resolver the tunnel uses.
    ///
    /// Stored by name, read back through `PrivateDns.of`, so a name this build
    /// does not know reads as the default rather than as no resolver at all.
    public var dnsMode: PrivateDns {
        get {
            guard privateDnsEnabled else { return .standard }
            return PrivateDns.of(storage.string(Key.dns))
        }
        set { storage.setString(Key.dns, newValue.rawValue) }
    }

    /// Whether the tunnel chooses the config, rather than the person.
    ///
    /// On by default: someone who has just bought a plan has no way to know
    /// which gateway is best for their network, and the tunnel can measure it.
    public var automaticServer: Bool {
        get { storage.bool(Key.automatic, default: true) }
        set { storage.setBool(Key.automatic, newValue) }
    }

    /// ISO code of the country automatic selection is limited to, or nil for
    /// anywhere.
    public var country: String? {
        get { storage.string(Key.country) }
        set { storage.setString(Key.country, newValue) }
    }

    /// What the person said they use the VPN for, as a `Purpose` name. It
    /// changes how automatic selection weighs its signals, and nothing else.
    public var purposeName: String? {
        get { storage.string(Key.purpose) }
        set { storage.setString(Key.purpose, newValue) }
    }

    public var purpose: Purpose { Purpose.of(purposeName) }

    /// What this phone has learned about each gateway, encoded by
    /// `ConnectionMemory`.
    ///
    /// It holds host:port, counters and timings — no credential and no profile
    /// line — but which gateways a phone has been talking to is not public
    /// either, so it goes wherever the rest of this does.
    public var connectionMemory: String? {
        get { storage.string(Key.memory) }
        set { storage.setString(Key.memory, newValue) }
    }

    /// The control-plane address that last answered.
    ///
    /// Kept so a first entry that has gone dark costs one failed connection at
    /// launch rather than one on every request. It is a hint, not a setting:
    /// the build's list decides what may be used, and this only reorders it.
    public var controlPlaneBase: String? {
        get { storage.string(Key.controlBase) }
        set { storage.setString(Key.controlBase, newValue) }
    }

    /// Signs the account out.
    ///
    /// Everything the account owns goes; the configs the person pasted in stay.
    /// Those were bought from somebody else, or handed over in a channel, and
    /// signing out of this app is not a reason to lose them — a person who had
    /// to re-paste forty configs to log back in would not log back in.
    public func clear() {
        let keepImported = importedConfigs
        storage.removeAll()
        importedConfigs = keepImported
    }
}

File: android/app/src/main/java/pro/syxvpn/app/data/SessionStore.kt
package pro.syxvpn.app.data

import android.content.Context
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import pro.syxvpn.app.BuildConfig
import pro.syxvpn.app.core.PrivateDns

/**
 * Holds the customer session token and the last known subscription URL.
 *
 * Both are bearer secrets, so they live in EncryptedSharedPreferences rather
 * than plain preferences: on a rooted or backed-up device a plain file is
 * readable, and the subscription URL alone is enough to use the account.
 */
class SessionStore(context: Context) {

    private val prefs = EncryptedSharedPreferences.create(
        context,
        "cvpn.session",
        MasterKey.Builder(context).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build(),
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
    )

    var token: String?
        get() = prefs.getString(KEY_TOKEN, null)
        set(value) = prefs.edit().apply {
            if (value == null) remove(KEY_TOKEN) else putString(KEY_TOKEN, value)
        }.apply()

    /** Whether there is a session at all. A fresh install has none, and that is fine. */
    val signedIn: Boolean get() = token != null

    var email: String?
        get() = prefs.getString(KEY_EMAIL, null)
        set(value) = prefs.edit().putString(KEY_EMAIL, value).apply()

    var subscriptionUrl: String?
        get() = prefs.getString(KEY_SUB_URL, null)
        set(value) = prefs.edit().putString(KEY_SUB_URL, value).apply()

    /** Last profile list fetched from the subscription, newline separated. */
    var cachedProfiles: String?
        get() = prefs.getString(KEY_PROFILES, null)
        set(value) = prefs.edit().putString(KEY_PROFILES, value).apply()

    /**
     * Configs the person pasted in themselves, newline separated.
     *
     * These are not a subscription and are nobody's to revoke: a config bought
     * from another provider belongs to whoever bought it, so it survives
     * signing out, an expired plan, and never having had one. Only removing it
     * here removes it.
     */
    var importedConfigs: String?
        get() = prefs.getString(KEY_IMPORTED, null)
        set(value) = prefs.edit().putString(KEY_IMPORTED, value).apply()

    /** The pasted configs as lines, in the order they were added. */
    val importedLines: List<String>
        get() = importedConfigs.orEmpty().split('\n').map { it.trim() }.filter { it.isNotEmpty() }

    fun addImported(uris: List<String>) {
        if (uris.isEmpty()) return
        importedConfigs = (importedLines + uris).joinToString("\n")
    }

    fun removeImported(uri: String) {
        importedConfigs = importedLines.filterNot { it == uri }.joinToString("\n").ifEmpty { null }
    }

    /**
     * Configs the person chose to hide, as the config lines themselves.
     *
     * A subscription decides which configs exist, so hiding is local and
     * sticky: a refresh brings one back from the control plane but it stays out
     * of the list until it is unhidden here.
     *
     * By the line, not by "host:port", which is what this held before. Two
     * configs into one gateway share an address, so hiding either of them hid
     * both — and on a subscription whose configs all point at one gateway, that
     * emptied the list.
     *
     * Stored as one string rather than a string set. The set went through
     * `putStringSet`, whose returned instance is the one the preferences hold,
     * and it did not reliably survive a restart; the pasted configs beside it
     * have always been a newline-joined string and have never lost anything.
     * The key is new, so nothing reads the old address-shaped entries: they
     * cannot be matched against a config line, and a config hidden under the
     * old scheme comes back once rather than staying hidden by an entry
     * nothing can ever clear.
     */
    var hiddenConfigs: Set<String>
        get() = prefs.getString(KEY_HIDDEN_LINES, null).orEmpty()
            .split('\n').map { it.trim() }.filter { it.isNotEmpty() }.toSet()
        set(value) = prefs.edit()
            .putString(KEY_HIDDEN_LINES, value.joinToString("\n").ifEmpty { null })
            .apply()

    /**
     * Whether the tunnel chooses the config, rather than the person.
     *
     * The Configs tab's setting. The VPN tab has no manual mode to store — it
     * never names a server — so nothing there reads or writes this.
     *
     * On by default: someone who has just bought a plan has no way to know
     * which gateway is best for their network, and the tunnel can measure it.
     * Choosing a config by hand turns it off, and it stays off until they pick
     * the Automatic row again.
     */
    /**
     * Which resolver the tunnel uses.
     *
     * Stored by name, read back through [PrivateDns.of], so a name this build
     * does not know reads as the default rather than as no resolver at all.
     *
     * `PRIVATE_DNS` gates it outright rather than only setting the default: a
     * build with the flag off has to behave exactly like the build before the
     * feature, including for the phone of somebody who had already picked an
     * encrypted mode. That is what makes it a switch that can be thrown back.
     */
    var dnsMode: PrivateDns
        get() = if (!BuildConfig.PRIVATE_DNS) {
            PrivateDns.STANDARD
        } else {
            PrivateDns.of(prefs.getString(KEY_DNS, null))
        }
        set(value) = prefs.edit().putString(KEY_DNS, value.name).apply()

    var automaticServer: Boolean
        get() = prefs.getBoolean(KEY_AUTOMATIC, true)
        set(value) = prefs.edit().putBoolean(KEY_AUTOMATIC, value).apply()

    /**
     * ISO code of the country automatic selection is limited to, or null for
     * anywhere. Set from the VPN screen's country list.
     */
    var country: String?
        get() = prefs.getString(KEY_COUNTRY, null)
        set(value) = prefs.edit().apply {
            if (value == null) remove(KEY_COUNTRY) else putString(KEY_COUNTRY, value)
        }.apply()

    /**
     * What the person said they use the VPN for, as a [pro.syxvpn.app.core.Purpose]
     * name. It changes how automatic selection weighs its signals, and nothing
     * else; the Configs tab never reads it.
     */
    var purposeName: String?
        get() = prefs.getString(KEY_PURPOSE, null)
        set(value) = prefs.edit().apply {
            if (value == null) remove(KEY_PURPOSE) else putString(KEY_PURPOSE, value)
        }.apply()

    /**
     * What this phone has learned about each gateway, encoded by
     * [pro.syxvpn.app.core.ConnectionMemory].
     *
     * It holds host:port, counters and timings — no credential and no profile
     * line — but it lives in the encrypted store with everything else, because
     * which gateways a phone has been talking to is not public either.
     *
     * Both the tunnel and the Configs health sweep write it. They read, change
     * and write the whole blob, so two writes in the same instant can lose a
     * measurement. That is the intended trade: the cost is one sample, and the
     * alternative is a lock held across a file write on the connect path.
     */
    var connectionMemory: String?
        get() = prefs.getString(KEY_MEMORY, null)
        set(value) = prefs.edit().apply {
            if (value == null) remove(KEY_MEMORY) else putString(KEY_MEMORY, value)
        }.apply()

    /**
     * The control-plane address that last answered.
     *
     * Kept so a first entry that has gone dark costs one failed connection at
     * launch rather than one on every request. It is a hint, not a setting:
     * the build's list decides what may be used, and this only reorders it.
     */
    var controlPlaneBase: String?
        get() = prefs.getString(KEY_CONTROL_BASE, null)
        set(value) = prefs.edit().apply {
            if (value == null) remove(KEY_CONTROL_BASE) else putString(KEY_CONTROL_BASE, value)
        }.apply()

    /**
     * Signs the account out.
     *
     * Everything the account owns goes; the configs the person pasted in stay.
     * Those were bought from somebody else, or handed over in a channel, and
     * signing out of this app is not a reason to lose them — a person who had
     * to re-paste forty configs to log back in would not log back in.
     */
    fun clear() {
        val keepImported = importedConfigs
        prefs.edit().clear().apply()
        importedConfigs = keepImported
    }

    private companion object {
        const val KEY_TOKEN = "token"
        const val KEY_EMAIL = "email"
        const val KEY_SUB_URL = "subscription_url"
        const val KEY_PROFILES = "profiles"
        const val KEY_HIDDEN_LINES = "hidden_config_lines"
        const val KEY_IMPORTED = "imported_configs"
        const val KEY_AUTOMATIC = "automatic_server"
        const val KEY_COUNTRY = "country"
        const val KEY_PURPOSE = "purpose"
        const val KEY_MEMORY = "connection_memory"
        const val KEY_CONTROL_BASE = "control_plane_base"
        const val KEY_DNS = "dns_mode"
    }
}
